### ISSUE TITLE
Loading a distributed checkpoint with Fabric fails with a RuntimeError

### ISSUE THREAD

**Issue by rasbt** [Lightning-AI/lightning#18853]
### Bug description

When using the Fabric `save` and `load` features in Lightning 2.1, it fails with a RuntimeError bug. I will post the long code files below, but in short, the gist is:

1) train a basic PyTorch model in `main-code.py`:

```python
    fabric = Fabric(accelerator="cuda", precision="bf16-mixed", devices=4, strategy="fsdp")
    fabric.launch()

    model, optimizer = fabric.setup(model, optimizer)

    # train model
    # then save model
    state = {
        "model": model, 
        "optimizer": optimizer,
        "anything-else-you-want-to-save": 123
    }

    fabric.save("checkpoint.ckpt", state)
```

2) Load model in `main-load.py`
```python
    fabric = Fabric(accelerator="cuda", precision="bf16-mixed", devices=4, strategy="fsdp")
    fabric.launch()

    model, optimizer = fabric.setup(model, optimizer)

    # train model
    # then save model
    state = {
        "model": model, 
        "optimizer": optimizer,
        "anything-else-you-want-to-save": 123
    }

    fabric.save("checkpoint.ckpt", state)
```

This results in the following error:

```
Traceback (most recent call last):
  File "/home/sebastian/Developer/cvpr2023/main-load.py", line 34, in <module>
Traceback (most recent call last):
Traceback (most recent call last):
  File "/home/sebastian/Developer/cvpr2023/main-load.py", line 34, in <module>
  File "/home/sebastian/Developer/cvpr2023/main-load.py", line 34, in <module>
    fabric.load("checkpoint.ckpt", state)
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/lightning/fabric/fabric.py", line 764, in load
    fabric.load("checkpoint.ckpt", state)
    fabric.load("checkpoint.ckpt", state)
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/lightning/fabric/fabric.py", line 764, in load
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/lightning/fabric/fabric.py", line 764, in load
    remainder = self._strategy.load_checkpoint(path=path, state=unwrapped_state, strict=strict)
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/lightning/fabric/strategies/fsdp.py", line 569, in load_checkpoint
    optim_state = load_sharded_optimizer_state_dict(
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/torch/distributed/checkpoint/optimizer.py", line 294, in load_sharded_optimizer_state_dict
    remainder = self._strategy.load_checkpoint(path=path, state=unwrapped_state, strict=strict)
    remainder = self._strategy.load_checkpoint(path=path, state=unwrapped_state, strict=strict)
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/lightning/fabric/strategies/fsdp.py", line 569, in load_checkpoint
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/lightning/fabric/strategies/fsdp.py", line 569, in load_checkpoint
    optim_state = load_sharded_optimizer_state_dict(
    optim_state = load_sharded_optimizer_state_dict(
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/torch/distributed/checkpoint/optimizer.py", line 294, in load_sharded_optimizer_state_dict
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/torch/distributed/checkpoint/optimizer.py", line 294, in load_sharded_optimizer_state_dict
    state_dict[key] = _shard_tensor(
    state_dict[key] = _shard_tensor(
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/torch/distributed/_shard/api.py", line 68, in _shard_tensor
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/torch/distributed/_shard/api.py", line 68, in _shard_tensor
    state_dict[key] = _shard_tensor(
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/torch/distributed/_shard/api.py", line 68, in _shard_tensor
    st = sharding_spec.shard(tensor, src_rank=src_rank, process_group=process_group)
    st = sharding_spec.shard(tensor, src_rank=src_rank, process_group=process_group)
    st = sharding_spec.shard(tensor, src_rank=src_rank, process_group=process_group)
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py", line 172, in shard
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py", line 172, in shard
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py", line 172, in shard
    assert local_tensor is not None
    assert local_tensor is not None
    assert local_tensor is not None
AssertionError
AssertionError
AssertionError
Traceback (most recent call last):
  File "/home/sebastian/Developer/cvpr2023/main-load.py", line 34, in <module>
    fabric.load("checkpoint.ckpt", state)
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/lightning/fabric/fabric.py", line 764, in load
    remainder = self._strategy.load_checkpoint(path=path, state=unwrapped_state, strict=strict)
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/lightning/fabric/strategies/fsdp.py", line 569, in load_checkpoint
    optim_state = load_sharded_optimizer_state_dict(
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/torch/distributed/checkpoint/optimizer.py", line 294, in load_sharded_optimizer_state_dict
    state_dict[key] = _shard_tensor(
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/torch/distributed/_shard/api.py", line 68, in _shard_tensor
    st = sharding_spec.shard(tensor, src_rank=src_rank, process_group=process_group)
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py", line 181, in shard
    dist.scatter(
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/torch/distributed/c10d_logger.py", line 47, in wrapper
    return func(*args, **kwargs)
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/torch/distributed/distributed_c10d.py", line 3139, in scatter
    _check_tensor_list(scatter_list, "scatter_list")
  File "/home/sebastian/miniforge3/envs/new-lit/lib/python3.9/site-packages/torch/distributed/distributed_c10d.py", line 807, in _check_tensor_list
    raise RuntimeError(
RuntimeError: Invalid function argument. Expected parameter `scatter_list` to be of type List[torch.Tensor].
```

I am attaching code so you can reproduce this. In the attached zip archive:

```
pip install -r requirements.txt
python main-code.py # only takes 1 minute
python main-load.py
```
[Archive.zip](https://github.com/Lightning-AI/lightning/files/13120412/Archive.zip)

### What version are you seeing the problem on?

master

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

cc @carmocca @justusschock @awaelchli

**Comment by rasbt** [comment]
The attached code file may have torch 2.0.1 in the `requirements.txt` and may not call `fabric.setup_dataloaders` in the loading script. But the issue still persists when upgrading to torch 2.1 and adding `fabric.setup_dataloaders`. 

Also, it works flawlessly when creating a single checkpoint file via 

```python
    strategy = FSDPStrategy(state_dict_type="full")
    fabric = Fabric(accelerator="cuda", precision="bf16-mixed", devices=4, strategy=strategy)
```

**Comment by awaelchli** [comment]
This was an issue in torch, and was fixed in https://github.com/pytorch/pytorch/pull/108915 if I read things correctly. The commit was released in 2.1.1. So the solution here is to upgrade `pip install -U torch`.
